tst/testlibgap/common.c

```diff
@@ -6,16 +6,20 @@
 void test_eval(const char * cmd)
 {
     Obj  res, ires;
-    Int  rc, i;
+    Int  rc, i, ok;
     printf("gap> %s\n", cmd);
-    res = GAP_EvalString(cmd);
-    rc = GAP_LenList(res);
-    for (i = 1; i <= rc; i++) {
-        ires = GAP_ElmList(res, i);
-        if (GAP_ElmList(ires, 1) == GAP_True) {
-            Char * buffer = GAP_CSTR_STRING(GAP_ElmList(ires, 5));
-            if (buffer)
-                printf("%s\n", buffer);
+    ok = GAP_Enter();
+    if (ok) {
```

Comment:
Yes, you might be right.  Do you even know off the top of your head (I'm not sure) if there is *any* guarantee whatsoever that there is any relation between the order some local variables are declared in C code, and what order they are placed on the stack (of course it could use registers for some of them too in which case it's irrelevant)?  AFAIK the C standard doesn't even know anything about a "stack" as that's a platform-specific implementation detail. But realistically speaking this discussion is only relevant where there is one.

So I think there's only so much we can do about this--it's just a best guess but not fool-proof.

Reply:
I think we already talked about exposing `CollectBags` in #3030.

It's not exactly clear to me what you want me to do here though.  Just manually test this case, or actually update the test to test this case in general somehow (e.g. have a copy of these tests that are compiled in such a way that forces the fallback)?

Reply:
The C standard definately puts no requirements on the order of variables on the stack, and will mix them up however it likes. If you are capturing a reference to a local variable in your fallback, the old remotely safe option would be to call GAP_enter, and then call a non-inlined function to actually do the work.

I'd probably just not put in a fallback which captures a reference to a local variable.

Reply:
To be clear, I'm 100% positive that fallback is going to lead to hard-to-track and horrible crashes, if there are any GAP variables in the current function scope, or in any inlined functions.

Reply:
> If you are capturing a reference to a local variable in your fallback, the old remotely safe option would be to call GAP_enter, and then call a non-inlined function to actually do the work.

That's what I was afraid of.  According to [Bill Allombert](https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1649776.html), "Nowadays about any compiler define `__GNUC__`", though no citation is given. But if so there's not much point in providing such a fallback anyways.

So maybe it would just be best, for the sake of moving forward, to `#error` for now and only worry about it if someone presents a case where this doesn't work for them.  Providing platform-specific fallbacks is easy enough, but I can't take the time now to provide fallbacks for every imaginable architecture where this might be used.

Reply:
Note that in *practice* most compilers do seem to allocate locals in order of definition on the stack (if they put them on the stack at all, i.e., an optimization could turn a variable into register only; but taking an address of it and passing it to a non-inline function *should* avoid that). So, adjusting the example to move all those local variables holding object references to be in between the enter/leave statements should effectively cure the problem *in that test* and *with (some) current compilers*. But I fully agree that this is (if it is even true, I did not test it) a very unstable situation, and could change and break at any time, which then indeed would lead to issues which fiendishly difficult to debug.

So indeed, it seems best to remove the generic fallback, and wait for people to complain that their favorite compiler does not work anymore with it.

(That said, of course in theory a compiler could put a local in a location which is neither a register or a stack; or it could put it there, but encrypt it; indeed, there are some plans in this direction as a form of mitigation against various security problems; any of these would break us; but they would also break GAP by itself. If this ever happens, we'll have to go back to the drawing board anyway (and e.g. switch to a compiler not doing that...)

Reply:
Oh, and of course we don't expect you, @embray , to provide fallbacks fo "every imaginable architecture". My stance on that is that it's better to wait for somebody who actually uses those architectures anyway.

Reply:
> Note that in practice most compilers do seem to allocate locals in order of definition on the stack

I have noted this as well, at least with gcc.  That's why I thought it was an "acceptable" fallback in the first place.

> but taking an address of it and passing it to a non-inline function should avoid that

It does, again at least in the case of gcc (and actually I can't think of any other reasonable thing for any compiler to do in that case: how are you supposed to pass a pointer to a value in a register?)  And of course, for locals that the compiler puts in registers this is a moot point (GASMAN handles that case separately with its setjmp hack).

> So indeed, it seems best to remove the generic fallback, and wait for people to complain that their favorite compiler does not work anymore with it.

+1  I don't like it, but it does seem like the safest option for now, alas.

Reply:
FWIW Boehm GC does use a similar fallback (they also declare the dummy variable volatile, which I intended to do as well but forgot): https://github.com/ivmai/bdwgc/blob/c50de12ab045b57953e152545836c23f76e7bc24/mark_rts.c#L502 
It also implements a [`GC_get_stack_base()`](https://github.com/ivmai/bdwgc/blob/1c62b365b088c409c7ee3546783194c8fdbef31b/os_dep.c#L793) for quite a number of different platforms, that I think is also supposed to work properly on per-thread stcks.

> they would also break GAP by itself. If this ever happens, we'll have to go back to the drawing board anyway

I think that may be the case anyways.  Is there really an advantage to continuing to maintain a separate GC when something more heavily used and robust like Boehm can be used?  Or at least, maybe, a customized fork thereof if there's really some fine-tuning necessary?

Reply:
gasman is much faster than Boehm, and various people's attempt to tune it haven't closed the gap.